openupgrade_scripts/scripts/account/15.0.1.1/pre-migration.py

```diff
@@ -0,0 +1,429 @@
+                    JOIN account_account_type aat ON aat.id = aa.user_type_id
+                    WHERE aml.move_id = am.id
+                    AND aat.type IN ('receivable', 'payable')) > 0
+                THEN false
+                ELSE true
+            END
+        WHERE am.always_tax_exigible IS NULL""",
+    )
+    # 2. Set always_tax_exigible = False
+    #    if record._collect_tax_cash_basis_values() is True
+    # 2.3 Set always_tax_exigible = False
+    #     if any(line.tax_line_id.tax_exigibility == 'on_payment')
+    #     else True
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_move am
+        SET always_tax_exigible =
+            CASE
+                WHEN (
+                    SELECT COUNT(*)
+                    FROM account_move_line aml
+                    JOIN account_tax tax ON tax.tax_exigibility = 'on_payment'
+                        AND aml.tax_line_id = tax.id
+                    WHERE aml.move_id = am.id) > 0
+                THEN false
+                ELSE true
+            END
+        WHERE am.always_tax_exigible IS NULL""",
+    )
+    # 2. Set always_tax_exigible = False
+    #    if record._collect_tax_cash_basis_values() is True
+    # 2.4 Set always_tax_exigible = False
+    #     if any(line.tax_ids.tax_exigibility == 'on_payment')
+    #     else True
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_move am
+        SET always_tax_exigible =
+            CASE WHEN (
+                SELECT COUNT(aml_tax_rel.account_tax_id)
+                FROM account_move_line_account_tax_rel aml_tax_rel
+                JOIN account_move_line aml ON aml.id = aml_tax_rel.account_move_line_id
+                    AND aml.move_id = am.id
+                JOIN account_tax tax ON tax.tax_exigibility = 'on_payment'
+                    AND tax.id = aml_tax_rel.account_tax_id) > 0
+            THEN false
+            ELSE true
+            END
+        WHERE am.always_tax_exigible IS NULL""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_move
+        SET always_tax_exigible = false
+        WHERE always_tax_exigible IS NULL""",
+    )
+
+
+def _fast_fill_account_move_amount_total_in_currency_signed(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE account_move
+        ADD COLUMN IF NOT EXISTS amount_total_in_currency_signed NUMERIC""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_move
+        SET amount_total_in_currency_signed =
+            CASE
+                WHEN move_type = 'entry'
+                THEN ABS(amount_total)
+                WHEN move_type IN ('in_invoice', 'out_refund', 'in_receipt')
+                THEN -amount_total
+                ELSE amount_total
+                END""",
+    )
+
+
+def _fast_fill_account_move_line_tax_tag_invert(env):
```

Comment:
The queries in this function are using too much subqueries and should be improved.